### PR TITLE
TF XLA greedy generation

### PR DESCRIPTION
**Pull request by Rocketknight1**
This is a first attempt at refactoring greedy generation in TF to use XLA. It has a couple of main problems still:

- Very hardcoded for GPT-2 right now
- No stopping based on EOS tokens yet (EDIT: Fixed!)
- Some really ugly array update hackery caused by the lack of tools like `dynamic_update_slice` (EDIT: Fixed!)
- `prepare_inputs_for_generation` assumes greedy generation, this is wrong (EDIT: Fixed)
- Outputs are slightly wrong compared to a manual loop - I'm still investigating (EDIT: Fixed!)

However, the outputs are coherent and the speedup compared to the old method is approximately 25X.

**UPDATE (from Patrick)**

PR should be good to go. See description below.

### CODE REVIEW

File: src/transformers/generation_tf_utils.py [hunk 1]
```diff
@@ -2195,6 +2204,45 @@ def greedy_search(
...
+        next_tokens = model_inputs["next_tokens"]
+        transposed_attention_mask = model_inputs["transposed_attention_mask"]
+        batch_size, seq_length = input_ids.shape
+
+        ptr = tf.ones(shape=(1,), dtype=tf.int32)
+        max_tokens_to_generate = max_length - seq_length
+
+        def greedy_search_cond_fn(input_ids, transposed_attention_mask, next_tokens, past, ptr):
+            """state termination condition fn."""
+            return ptr < max_tokens_to_generate
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2221]: Can you check here in a hacky way whether there is an EOS in `input_ids` to stop?

**Comment by Rocketknight1** [comment]
Outputs now correctly match a manual loop, up to minor float32 errors. Code has been simplified with the addition of direct calls to XLA, and we're now at >30X speedup versus the original code.

File: src/transformers/generation_tf_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py]: Renamed to `current_pos`!

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -2195,6 +2206,52 @@ def greedy_search(
...
+        current_pos = tf.ones(shape=(1,), dtype=tf.int32)
+        max_tokens_to_generate = max_length - seq_length
+
+        def greedy_search_cond_fn(generated, finished_sequences, attention_mask, next_tokens, past, current_pos):
+            """state termination condition fn."""
+            return ~tf.reduce_all(finished_sequences)
+
+        def greedy_search_body_fn(generated, finished_sequences, attention_mask, next_tokens, past, current_pos):
+            """state update fn."""
+            position_ids = tf.reduce_sum(attention_mask, axis=1, keepdims=True) - 1
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2232]:
```suggestion
    if xla_compile:
            position_ids = tf.reduce_sum(attention_mask, axis=1, keepdims=True) - 1
            model_inputs = {"input_ids": next_tokens, "past": past, "attention_mask": attention_mask, position_ids=position_ids}
   else: 
            input_ids = generated[:, :curres_pos]
            # prep attention_mask
            # prep encoder_decoder
            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs) 
```

**Comment by patrickvonplaten** [comment]
@Rocketknight1 

I've been testing the code by running the following code example - put it in a `run.py` file:

```python
from transformers import TFGPT2LMHeadModel, GPT2Tokenizer
import tensorflow as tf
import sys

use_xla = bool(int(sys.argv[1]))

physical_devices = tf.config.list_physical_devices('GPU')
for device in physical_devices:
    tf.config.experimental.set_memory_growth(device, True)

tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
model = TFGPT2LMHeadModel.from_pretrained("gpt2")

input_ids = tokenizer("hello there can you continue", return_tensors="tf").input_ids

outputs = model.generate(input_ids, use_xla=use_xla)

print("Output", tokenizer.batch_decode(outputs))
```

XLA-case `python run.py 1`
Non-XLA-case `python run.py 0`

File: src/transformers/generation_tf_utils.py [hunk 3]
```diff
@@ -2195,6 +2231,86 @@ def greedy_search(
...
+
+                model_outputs = self(input_ids=next_tokens, past=past,
+                                     attention_mask=attention_mask, position_ids=position_ids)
+                logits = model_outputs.logits[:, -1:]
+                next_tokens = tf.argmax(logits, axis=-1, output_type=tf.int32)
+                finished_sequences = finished_sequences | (next_tokens[:, 0] == eos_token_id)
+                past = model_outputs.past_key_values
+                new_past_index = current_pos + seq_length - 1
+
+                if use_xla:
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2292]: Can be switched on/off - `use_xla` is an input flag to `generate()`

**Comment by patrickvonplaten** [comment]
Design is now more or less fixed. Running now all important slow tests: 

```python
RUN_SLOW=1 pytest -sv \
tests/bart/test_modeling_tf_bart.py \
tests/t5/test_modeling_tf_t5.py \
tests/gpt2/test_modeling_tf_gpt2.py \
tests/vision_encoder_decoder/test_modeling_tf_vision_encoder_decoder.py \
tests/encoder_decoder/test_modeling_tf_encoder_decoder.py \
tests/speech_to_text/test_modeling_tf_speech_to_text.py
```

as well as

```bash
RUN_SLOW=1 pytest -sv tests/test_modeling_tf_rag.py
```
(this one is tested on brutasse as it requires a heavy index).

File: tests/gpt2/test_modeling_tf_gpt2.py [hunk 4]
```diff
@@ -513,3 +530,16 @@ def test_lm_generate_gpt2(self):
         # fmt: on
         output_ids = model.generate(input_ids, do_sample=False)
         self.assertListEqual(output_ids[0].numpy().tolist(), expected_output_ids)
+
+    @slow
+    def test_lm_generate_gpt2_xla(self):
```
Comment (patrickvonplaten) [inline tests/gpt2/test_modeling_tf_gpt2.py:535]: This test verifies that XLA works exactly like non-XLA

**Comment by patrickvonplaten** [comment]
The PR is now ready for merge - we now have a XLA-compileable `greedy_search` method which fully replaced the old `greedy_search` generate method. The XLA-compilable `greedy_search` behaves exactly the same way as the previous one for eager execution, but also allows for XLA compilation. 
This was made possible at the expense of some code overhead (see function decorator flags), and a couple of `if use_xla` statements.

Since XLA-compileable `generate` has model-dependent cache tensors, we need to add a **single** new function to every model we want to make XLA-compilable, called `_update_model_kwargs_for_xla_generation`. One is added for GPT2 now so that GPT2 now works with XLA. The method is a bit hacky at the moment and could definitely be cleaned up in a future PR - I've copy-pasted some code from the previous state of the PR. XLA gpt2 generate gives the same result as non-XLA generate though.

IMO, the design is more or less fixed now as we have 0 breaking changes, no duplicated (non-XLA and XLA) generate methods, and relatively clean code.

Some important things that should be discussed here though:

1 - At the moment the way the cache and attention_mask is padded to make it XLA-compatible is a bit hacky (see code in `_update_model_kwargs_for_xla_generation`), but it allows for 0 breaking changes. It might be preferable to change the way `past_key_values` is implemented inside the model to potentially make generate even faster and less "hacky". This would mean however that `past_key_values` would change its tensor shape which is a breaking change though since we do return `past_key_values` when one passes `use_cache=True`. So might not be worth it

2. - XLA can be enabled with a flag `use_xla=True/False` which defaults to `False` at the moment. Once we make all models XLA compileable we could default the flag to `True`. TF has a somewhat weird way of enabling jit-XLA compilation with the function decorator which is why we have so many functions inside functions now in this PR. I do think it's the most intuitive way though is to allow the user to pass a `use_xla=True/False` flag. But I'm no TF expert - what do you think @sgugger @Rocketknight1 @gante ?

**Comment by patrickvonplaten** [comment]
Might be a bit difficult to review the git diff here. Would be nice if @Rocketknight1 and @gante, you could go through the new file and carefully check if everything makes sense - happy to answer any question you might have.

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 5]
```diff
@@ -851,15 +852,89 @@ def get_output_embeddings(self):
...
+
+        return {
+            "input_ids": inputs,
+            "attention_mask": attention_mask,
+            "position_ids": position_ids,
+            "past_key_values": past,
+            "use_cache": use_cache,
+        }
+
+    def _update_model_kwargs_for_xla_generation(self, outputs, model_kwargs, current_pos, max_length):
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:883]: We only need to add a single new method for every model that we want to make XLA-compatible

File: src/transformers/models/rag/modeling_tf_rag.py [hunk 6]
```diff
@@ -1309,9 +1309,14 @@ def extend_enc_output(tensor, num_beams=None):
...
-            # TODO(Patrick) clean-up once generate is fully cleaned up
             model_kwargs["attention_mask"] = context_attention_mask
-            # TODO(Patrick) remove once generate is fully cleaned up
+
+            if model_kwargs.get("encoder_attentions", None) is None:
+                model_kwargs.pop("encoder_attentions", None)
+            if model_kwargs.get("encoder_hidden_states", None) is None:
+                model_kwargs.pop("encoder_hidden_states", None)
+
+            model_kwargs.pop("output_hidden_states", None)
```
Comment (patrickvonplaten) [inline src/transformers/models/rag/modeling_tf_rag.py:1319]:
```suggestion
```

File: tests/gpt2/test_modeling_gpt2.py [hunk 7]
```diff
@@ -660,29 +660,16 @@ def _test_lm_generate_gpt2_helper(
...
         model.to(torch_device)
-        input_ids = torch.tensor([[464, 3290]], dtype=torch.long, device=torch_device)  # The dog
+
+        # The dog
+        input_ids = torch.tensor([[464, 3290]], dtype=torch.long, device=torch_device)
+
+        # The dog was found in a field near the intersection of West and West Streets.\n\nThe dog
+        # fmt: off
         expected_output_ids = [
-            464,
```
Comment (patrickvonplaten) [inline tests/gpt2/test_modeling_gpt2.py:665]: just a clean-up

File: tests/gpt2/test_modeling_tf_gpt2.py [hunk 8]
```diff
@@ -294,6 +294,19 @@ def create_and_check_gpt2_lm_head(self, config, input_ids, input_mask, head_mask
         result = model(inputs)
         self.parent.assertEqual(result.logits.shape, (self.batch_size, self.seq_length, self.vocab_size))
 
+    def create_and_check_gpt2_xla_generate(self, config, *args):
```
Comment (patrickvonplaten) [inline tests/gpt2/test_modeling_tf_gpt2.py:297]: check that xla and non-xla is always the same. Using the same input_ids here since it's a bit flaky (1/20 runs there seem to be some numerical small differences) and because XLA doesn't cut output ids in case EOS token is generated.

File: src/transformers/generation_tf_utils.py [hunk 9]
```diff
@@ -16,6 +16,7 @@
 
 import inspect
 from dataclasses import dataclass
+from functools import wraps
```
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:19]: Nit: I'd prefer to just import `functools` and then call `functools.wraps` later on to make it clearer where the `wraps` comes from.

File: src/transformers/generation_tf_utils.py [hunk 10]
```diff
@@ -1717,6 +1742,13 @@ def _update_model_kwargs_for_generation(
 
         return model_kwargs
 
+    def _update_model_kwargs_for_xla_generation(
+        self, outputs: ModelOutput, model_kwargs: Dict[str, Any], current_pos: tf.Tensor, max_length: int
+    ) -> Dict[str, Any]:
+        raise NotImplementedError(
+            f"{self.__class__} is not compileable with XLA at the moment. You should implement a `_update_model_kwargs_for_xla_generation` in the respective modeling file for XLA-compatible generation."
+        )
```
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:1750]: Can it be shorter to fi the 119 char limit :-)

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:883]: Can the bulk of it be in a base method on TFPreTrainedModel?

**Review by sgugger — APPROVED** [review]
No problem with the design! I wonder if we want to have `use_xla` defaults to something in the config, like all the other variables of `generate`?

Thanks a lot for working on this @patrickvonplaten !

File: src/transformers/models/gpt2/modeling_tf_gpt2.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/gpt2/modeling_tf_gpt2.py:883]: Yeah definitely! This function can be cleaned up and I think we can really restrict the amount of model-specific code to a minimum here

**Comment by gante** [comment]
> 2 - XLA can be enabled with a flag use_xla=True/False which defaults to False at the moment. Once we make all models XLA compileable we could default the flag to True. TF has a somewhat weird way of enabling jit-XLA compilation with the function decorator which is why we have so many functions inside functions now in this PR. I do think it's the most intuitive way though is to allow the user to pass a use_xla=True/False flag. But I'm no TF expert - what do you think @sgugger @Rocketknight1 @gante ?

There is an alternative solution -- using the `tf.config` functions to trigger whether we want [XLA](https://www.tensorflow.org/api_docs/python/tf/config/optimizer/set_jit) or [eager execution](https://www.tensorflow.org/api_docs/python/tf/config/run_functions_eagerly) on functions decorated with `tf.function`. This approach has pros and cons, with respect to the current one.

Pros:
1. No need to define inner functions with `@xla_compile`. Everything that might get XLA compilation gets a `@tf.function`, and the exec mode is set by calling the appropriate config at the beginning of `generate()`
2. Repeated `generate()` calls do not redefine the functions, so they should benefit from previous XLA warmup -- not sure if this is a significant advantage or not 🤔 

Cons:
1. Setting any of these `tf.config` methods requires an empty TF session. This means we can't call a generate without and with XLA without resetting the session. Also, if we were to call it at the beginning of `generate()`, no compiled TF model could exist (again, because it needs an empty session).

Personally, I'm slightly leaning towards this approach (with the `tf.config` calls) if we could make it work, but no strong feelings :)

**Comment by patrickvonplaten** [comment]
> > 2 - XLA can be enabled with a flag use_xla=True/False which defaults to False at the moment. Once we make all models XLA compileable we could default the flag to True. TF has a somewhat weird way of enabling jit-XLA compilation with the function decorator which is why we have so many functions inside functions now in this PR. I do think it's the most intuitive way though is to allow the user to pass a use_xla=True/False flag. But I'm no TF expert - what do you think @sgugger @Rocketknight1 @gante ?
> 
> There is an alternative solution -- using the `tf.config` functions to trigger whether we want [XLA](https://www.tensorflow.org/api_docs/python/tf/config/optimizer/set_jit) or [eager execution](https://www.tensorflow.org/api_docs/python/tf/config/run_functions_eagerly) on functions decorated with `tf.function`. This approach has pros and cons, with respect to the current one.
> 
> Pros:
> 
> 1. No need to define inner functions with `@xla_compile`. Everything that might get XLA compilation gets a `@tf.function`, and the exec mode is set by calling the appropriate config at the beginning of `generate()`
> 2. Repeated `generate()` calls do not redefine the functions, so they should benefit from previous XLA warmup -- not sure if this is a significant advantage or not thinking  (could be, for instance, on hosted APIs that expect repeated calls on the same function)
> 
> Cons:
> 
> 1. Setting any of these `tf.config` methods requires an empty TF session. This means we can't call a generate without and with XLA without resetting the session. Also, if we were to call it at the beginning of `generate()`, no compiled TF model could exist (again, because it needs an empty session).
> 
> Personally, I'm slightly leaning towards this approach (with the `tf.config` calls) if we could make it work, but no strong feelings :)

Happy to switch to using `tf.config(...)` if you guys prefer! I'm not familiar with what is intuitive in TF so trusting you guys here. But I'm also not too happy with all those subfunctions and complicated function wrappers, so this approach makes a lot of sense. 
Just to understand better, do you think it make more sense to set `tf.config.optimizer.set_jit(enabled=use_xla)` right after `generate(...)` or should we move it out completely of `generate()` and let the user set it before calling generate? In this case we wouldn't need a `use_xla` flag here at all no? Not sure what's better here

**Comment by gante** [comment]
> Just to understand better, do you think it make more sense to set tf.config.optimizer.set_jit(enabled=use_xla) right after generate(...) or should we move it out completely of generate() and let the user set it before calling generate? In this case we wouldn't need a use_xla flag here at all no? Not sure what's better here

There would be no need for the `use_xla` flag in `tf.function`. I'd say to keep it in `generate()`, as it should be the outermost function that may need XLA inside. @Rocketknight1 what do you think (current approach vs `tf.config`?)

**Review by gante — APPROVED** [review]
LGTM 👍

**Review by LysandreJik — APPROVED** [review]
Great work!

File: src/transformers/models/t5/modeling_tf_t5.py [hunk 11]
```diff
@@ -1560,11 +1563,76 @@ def prepare_inputs_for_generation(
             "past_key_values": past,
             "encoder_outputs": encoder_outputs,
             "attention_mask": attention_mask,
+            "decoder_attention_mask": decoder_attention_mask,
             "head_mask": head_mask,
             "decoder_head_mask": decoder_head_mask,
             "use_cache": use_cache,
         }
 
+    def _update_model_kwargs_for_xla_generation(self, outputs, model_kwargs, current_pos, max_length):
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_tf_t5.py:1572]: T5 works as well now

File: tests/t5/test_modeling_tf_t5.py [hunk 12]
```diff
@@ -454,6 +476,27 @@ def test_train_pipeline_custom_model(self):
...
 class TFT5GenerationIntegrationTests(unittest.TestCase):
+    @slow
+    def test_greedy_xla_generate_simple(self):
+        model = TFT5ForConditionalGeneration.from_pretrained("t5-small", min_length=-1)
+        tokenizer = T5Tokenizer.from_pretrained("t5-small")
+
+        sentence = "Translate English to German: Today is a beautiful day."
+        input_ids = tokenizer(sentence, return_tensors="tf", padding=True).input_ids
+
+        xla_generate = tf.function(model.generate, jit_compile=True)
```
Comment (patrickvonplaten) [inline tests/t5/test_modeling_tf_t5.py:487]: This is the only line one has to add now to make it xla_compile-able - works with T5 and GPT2 now. Ok for you @Rocketknight1 @gante ?

File: src/transformers/generation_utils.py [hunk 13]
```diff
@@ -733,7 +733,7 @@ def _get_logits_processor(
                 )
         if bad_words_ids is not None:
             processors.append(NoBadWordsLogitsProcessor(bad_words_ids, eos_token_id))
-        if min_length is not None and eos_token_id is not None and min_length > -1:
+        if min_length is not None and eos_token_id is not None and min_length > 0:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:736]:
```suggestion
        if min_length is not None and eos_token_id is not None and min_length > -1:
```
